### PR TITLE
refactor(mla): isolate planned FA2, FA3, and CUTLASS backends

### PR DESCRIPTION
**Pull request by saltyminty**
## 📌 Description

This is the first standalone section extracted from [the original Batch MLA redesign PR](https://github.com/flashinfer-ai/flashinfer/pull/4031). That PR covered several independent backend and API changes; this series breaks the work into smaller PRs that can be reviewed, validated, and landed individually.

### Current scope

This PR establishes the planned Batch MLA foundation for the existing FA2, FA3, and CUTLASS backends:

- moves concrete planning and execution state into backend-owned implementations;
- introduces canonical plan metadata and explicit packed/split query and KV-cache layout contracts;
- keeps the planned `run()` path backend-native, with no runtime backend discovery or replanning, and reuses plan-owned workspaces and empty-LSE storage;
- makes CUDA Graph replanning transactional while preserving the wrapper attributes used by SGLang's fast replay path;
- preserves legacy public imports, positional/flat-CSR planning, split tensor calls, dynamic LSE behavior, and caller-owned output/LSE identity;
- retains deprecated planless CUTLASS execution as a validated per-call path without publishing or mutating planned state;
- preserves historical trace identities while allowing trace replay to recognize the reorganized implementation; and
- documents the backend ownership and lifecycle conventions introduced by this slice.

The structural input resolver also preserves the existing zero-width PE compatibility case: a packed plan can still accept `(left, empty_right)` without requiring a copy.

### Incoming follow-up sections

Subsequent PRs are expected to cover, independently:

1. a public sync-free CUDA Graph plan-update API and removal of the temporary private SGLang bridge;
2. TRTLLM-GEN, XQA, and CuTe DSL backend verticals;
3. deterministic automatic selection with typed unsupported-backend fallback;
4. the unified tensor-first functional API; and
5. optional autotuning and benchmark follow-ups.

## 🔍 Related Issues

- Original umbrella PR: https://github.com/flashinfer-ai/flashinfer/pull/4031
- Related issue: https://github.com/flashinfer-ai/flashinfer/issues/4037

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

The complete repository-wide pre-commit suite passes.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Validation completed:

- planned-wrapper compatibility suite: 71 passed;
- focused planned CUTLASS/FP8 suite: 10 passed;
- trace and trace-apply suite: 26 passed;
- repository-wide pre-commit suite and `git diff --check` passed;
- targeted SM90 and SM100 real-GPU checks passed for FA2/FA3/CUTLASS numerical execution, CUDA Graph replay and replanning, legacy flat-CSR/separate-tensor/dynamic-LSE calls, FP8 CUTLASS execution, and zero-width PE packed-plan compatibility; and
- targeted SGLang SM100 integration passed fast decode/prefill planning, wrapper mirror identity, canonical/legacy numerical parity, focused MLA tests, and unified dense block-table tests.

Full repository tests, wheel/build validation, and the repository-wide GPU CI matrix have not been run locally.

## Reviewer Notes

The first two commits intentionally preserve the review boundary between the planned-contract foundation and the subsequent compatibility/hot-path corrections. The final one-file commit only gives the cached FA planning helper its concrete JIT-generator signature so repository-wide mypy can validate it; cache keys and runtime behavior are unchanged.

Suggested review focus:

- backend ownership and the plan/publication transaction boundary;
- planned `run()` hot-path work and workspace/LSE reuse;
- packed/split structural input compatibility, including zero-width PE;
- CUDA Graph replan behavior and SGLang mirror attributes;
- deprecated planless CUTLASS isolation; and
- legacy imports, call forms, output identity, and trace compatibility.

This branch is intentionally based on the validated extraction point from #4031 rather than rebased after validation; current-main CI should be treated as the integration check for intervening changes.

## Summary by CodeRabbit

* **New Features**
  * Added a planned Batch MLA paged-attention API supporting CSR, dense, and combined metadata.
  * Added packed and split query/KV inputs, FP8 output scaling, CUDA Graph support, and automatic FA2, FA3, or CUTLASS backend selection.
  * Added planless CUTLASS execution for supported configurations.

* **Compatibility**
  * Preserved legacy calling patterns with deprecation warnings.

* **Documentation**
  * Added comprehensive Batch MLA architecture, API, and backend documentation.

* **Bug Fixes**
  * Improved MLA tracing and validation for metadata, layouts, data types, scaling, and backend capabilities.

### CODE REVIEW

**Comment by saltyminty** [comment]
The GitHub-actions flagged API breaking changes are false positives – `BatchMLAPagedAttentionWrapper` was moved but is still re-exported from `flashinfer/mla/_core.py`.

File: docs/design_docs/batch_mla_backend_architecture.md [hunk 1]
```diff
@@ -0,0 +1,435 @@
+# Batch MLA backend architecture
+
+This document describes the backend boundary for
+`BatchMLAPagedAttentionWrapper` and its FA2, FA3, and CUTLASS backends. Public
+access is through `flashinfer.mla`; the `_batch_mla` package and its concrete
+backend classes are private implementation details.
```
[the remark this answers is not in the document]
Comment (saltyminty) [reply docs/design_docs/batch_mla_backend_architecture.md:6]: None of the existing docs in flashinfer use this convention

File: docs/design_docs/batch_mla_backend_architecture.md
Comment (bkryu) [inline docs/design_docs/batch_mla_backend_architecture.md:1]: Could be great to have example usages in the design doc.

While you add the example usage, can you be mindful of how it can extend/change as we consolidate all backends?

**Comment by qsang-nv** [comment]
### Three merge-blocking issues

**1. Planned CUTLASS loses the packed-input contiguity guarantee and lacks cross-device validation.**

`main` always went through `torch.cat([q_nope, q_pe], -1)`, which implicitly produced a contiguous tensor. The zero-copy structural path drops that: `_contracts.py:255-276` checks only `ndim == 3`, last dim, and dtype — no `is_contiguous()`, no device check (`cutlass_backend.py:270-271` validates the device of `kv_len`/`page_table` only). The kernel never reads tensor strides:

```cpp
// include/flashinfer/attention/cutlass_mla.cuh:97-100
StrideQ stride_Q = cute::make_tuple(int64_t(D_latent + D_rope), _1{},
                                    int64_t(H * (D_latent + D_rope)));
```

So `run(query=big[:, :, :576], ...)` passes validation and is silently misread — wrong numbers, no error. For device, `csrc/cutlass_mla.cu` does `CUDADeviceGuard(q_nope_pe.device().device_id)`, i.e. it follows the query, so a query on a device other than the workspace can make accesses through the workspace and page-table pointers invalid on the selected device.

The contiguity issue is a regression relative to `main`; the cross-device relationship was not fully validated on `main` either.

The *derived* packed path is already safe here — `_are_adjacent_last_dim_views` verifies row-major density before `as_strided`. Only the caller-supplied packed tensor is unchecked. FA2/FA3 are unaffected by the stride issue because `batch_mla_run.cu:48-57` forwards `stride(0)`/`stride(1)`.

**2. The pinned staging buffer is exposed to allocator-wide reuse.**

`_fa_common.py:130-140` allocates a fresh int workspace and pinned buffer per `plan()` unless one is passed in, and `_wrapper.py:532-541` only passes one for a CUDA-graph replan.

After a later successful non-graph replan replaces that backend, the previous backend can lose its last reference while its H2D copy — and a previously enqueued run ordered after that copy — are still pending. Because the raw `cudaMemcpyAsync` at `scheduler.cuh:1836` is invisible to PyTorch's caching host allocator, no reuse event is recorded. The freed block can therefore be handed to another pinned allocation and overwritten while the copy is still in flight. Corrupting the work-partition metadata copied into the integer workspace can produce wrong output or an OOB access with nothing on the stack pointing to the cause.

On `main`, the allocation remained owned by the wrapper, so this allocator-wide reuse window did not exist. That said, restoring wrapper ownership alone is not a complete synchronization guarantee: the staging storage must also remain unmodified until the previous asynchronous H2D has completed.

**3. `_retired_cuda_graph_backends` grows without bound.**

`_wrapper.py:643-648` appends the previous backend after every successful CUDA-graph replan, with no `pop`, `clear`, or cap anywhere in the implementation. When an integration calls `plan()` every decode step under CUDA graphs, one object is appended per step indefinitely. At 200 plans/s this retains 17.28 million backend objects per day. The large workspaces are shared, but the list entries, backend instances, and per-plan state remain uniquely retained, so memory growth is still unbounded. `main` had no per-plan backend objects at all.

File: docs/design_docs/batch_mla_backend_architecture.md
Comment (saltyminty) [reply docs/design_docs/batch_mla_backend_architecture.md:1]: See https://github.com/flashinfer-ai/flashinfer/pull/4697/commits/05ae7d6bf3ffb149a8afc3d9c700d4404a88e55d

**Comment by saltyminty** [comment]
@qsang-nv addressed in https://github.com/flashinfer-ai/flashinfer/pull/4697/commits/05ae7d6bf3ffb149a8afc3d9c700d4404a88e55d

**Review by qsang-nv — APPROVED** [review]
